File: src/QuantizedTessellation3d.cc
//------------------------------------------------------------------------
// QuantizedTessellation3d
//
// An intermediate representation for 3D tessellations in integer
// coordinates.
//------------------------------------------------------------------------
#include <iostream>
#include <algorithm>
#include <map>

#include "QuantizedTessellation3d.hh"
#include "polytope.hh"
#include "polytope_tessellator_utilities.hh"

namespace polytope {
  
//------------------------------------------------------------------------------
// Construct with the given generators.  Finds the bounding limits and sets the
// quantized generators.
//------------------------------------------------------------------------------
template<typename IntType, typename RealType>
QuantizedTessellation3d<IntType, RealType>::
QuantizedTessellation3d(const std::vector<RealType>& points,
                        const std::vector<RealType>& boundaryPoints) {
  geometry::computeBoundingBox<3, RealType>(&boundaryPoints[0], boundaryPoints.size(), true, xmin, xmax);
  length = std::max(xmax[0] - xmin[0], std::max(xmax[1] - xmin[1], xmax[2] - xmin[2]));
  xmin[0] -= 2.0*length;
  xmin[1] -= 2.0*length;
  xmin[2] -= 2.0*length;
  xmax[0] += 2.0*length;
  xmax[1] += 2.0*length;
  xmax[2] += 2.0*length;
  length *= 5.0;
  this->construct(points);
}

//------------------------------------------------------------------------------
// Construct with the given generators using the specified bounds.
//------------------------------------------------------------------------------
template<typename IntType, typename RealType>
QuantizedTessellation3d<IntType, RealType>::
QuantizedTessellation3d(const std::vector<RealType>& points,
                        const RealType xmin_in[3],
                        const RealType xmax_in[3]) {
  xmin[0] = xmin_in[0];
  xmin[1] = xmin_in[1];
  xmin[2] = xmin_in[2];
  xmax[0] = xmax_in[0];
  xmax[1] = xmax_in[1];
  xmax[2] = xmax_in[2];
  length = std::max(xmax[0] - xmin[0], xmax[1] - xmin[1]);
  this->construct(points);
}

//------------------------------------------------------------------------------
// Convert real coordinates to integers.
//------------------------------------------------------------------------------
template<typename IntType, typename RealType>
void
QuantizedTessellation3d<IntType, RealType>::
quantize(const RealType* realcoords, IntType* intcoords) const {
  const RealType dx = length/(coordMax - coordMin);
  intcoords[0] = coordMin + IntType((realcoords[0] - xmin[0])/dx);
  intcoords[1] = coordMin + IntType((realcoords[1] - xmin[1])/dx);
  intcoords[2] = coordMin + IntType((realcoords[2] - xmin[2])/dx);
}

//------------------------------------------------------------------------------
// Convert int coordinates to reals.
//------------------------------------------------------------------------------
template<typename IntType, typename RealType>
void
QuantizedTessellation3d<IntType, RealType>::
dequantize(const IntType* intcoords, RealType* realcoords) const {
  const RealType dx = length/(coordMax - coordMin);
  realcoords[0] = xmin[0] + (intcoords[0] - coordMin)*dx;
  realcoords[1] = xmin[1] + (intcoords[1] - coordMin)*dx;
  realcoords[2] = xmin[2] + (intcoords[2] - coordMin)*dx;
}

//------------------------------------------------------------------------------
// Read out the current quantized tessellation state to a regular tessellation.
//------------------------------------------------------------------------------
template<typename IntType, typename RealType>
void
QuantizedTessellation3d<IntType, RealType>::
fillTessellation(Tessellation<3, RealType>& mesh) const {
  POLY_ASSERT2(false, "Implement me!");
}

//------------------------------------------------------------------------------
// Internal method to construct once we have the bounds set.
// Assumes xmin, xmax, and length are already set.
//------------------------------------------------------------------------------
template<typename IntType, typename RealType>
void
QuantizedTessellation3d<IntType, RealType>::
construct(const std::vector<RealType>& points) {
  POLY_ASSERT(points.size() % 3 == 0);
  POLY_ASSERT(std::abs(xmax[0] - xmin[0] - length) < 1e-10*length);
  POLY_ASSERT(std::abs(xmax[1] - xmin[1] - length) < 1e-10*length);
  POLY_ASSERT(std::abs(xmax[2] - xmin[2] - length) < 1e-10*length);
  const int numGenerators = points.size()/3;
  generators.resize(numGenerators);
  for (unsigned i = 0; i < numGenerators; ++i) {
    this->quantize(&points[3*i], &generators[i].x);
    generators[i].index = i;
  }
  POLY_ASSERT2(false, "Implement 3D guard generators!");
}

//------------------------------------------------------------------------------
// Instantiate versions we know we need.
//------------------------------------------------------------------------------
template<typename IntType, typename RealType> 
IntType
QuantizedTessellation3d<IntType, RealType>::coordMin = std::numeric_limits<IntType>::min()/3;

template<typename IntType, typename RealType> 
IntType
QuantizedTessellation3d<IntType, RealType>::coordMax = std::numeric_limits<IntType>::max()/3;

template struct QuantizedTessellation3d<int, double>;
}

File: src/PYB11/QuantizedTessellation3d.py
from PYB11Generator import *

@PYB11template("IntType", "RealType")
class QuantizedTessellation2d:
    """QuantizedTessellation2d

An intermediate representation for 2D tessellations in integer
coordinates."""

    PYB11typedefs = """
  typedef Point3<%(IntType)s> IntPoint;
  typedef Point3<%(RealType)s> RealPoint;
"""

    #...........................................................................
    # Constructors
    def pyinit1(self,
                points = "const std::vector<%(RealType)s>&",
                boundaryPoints = "const std::vector<%(RealType)s>&"):
        "Construct with the given generators.  Finds the bounding limits and sets the quantized generators."

    @PYB11implementation("""[](const std::vector<%(RealType)s>& points,
                               py::tuple xmin_in,
                               py::tuple xmax_in) {
                                   %(RealType)s xmin[3] = {xmin_in[0].cast<%(RealType)s>(),
                                                           xmin_in[1].cast<%(RealType)s>(),
                                                           xmin_in[2].cast<%(RealType)s>()};
                                   %(RealType)s xmax[3] = {xmax_in[0].cast<%(RealType)s>(),
                                                           xmax_in[1].cast<%(RealType)s>(),
                                                           xmax_in[2].cast<%(RealType)s>()};
                                   return new QuantizedTessellation3d<%(IntType)s, %(RealType)s>(points, xmin, xmax);
                               }""")
    def pyinit2(self,
                points = "const std::vector<%(RealType)s>&",
                xmin = "py::tuple",
                xmax = "py::tuple"):
        "Construct with the given generators using the specified bounds."

    #...........................................................................
    # Methods
    @PYB11const
    @PYB11implementation("""[](const QuantizedTessellation3d<%(IntType)s, %(RealType)s>& self,
                               py::tuple& realcoords) {
                                   %(RealType)s rcoords[3] = {realcoords[0].cast<%(RealType)s>(),
                                                              realcoords[1].cast<%(RealType)s>(),
                                                              realcoords[2].cast<%(RealType)s>()};
                                   %(IntType)s icoords[3];
                                   self.quantize(rcoords, icoords);
                                   return py::make_tuple(icoords[0], icoords[1], icoords[2]);
                               }""")
    def quantize(self,
                 realcoords = "py::tuple"):
        "Convert real coordinates to integers."
        return "py::tuple"

    @PYB11const
    @PYB11implementation("""[](const QuantizedTessellation3d<%(IntType)s, %(RealType)s>& self,
                               py::tuple& intcoords) {
                                   %(IntType)s icoords[3] = {intcoords[0].cast<%(IntType)s>(),
                                                             intcoords[1].cast<%(IntType)s>(),
                                                             intcoords[2].cast<%(IntType)s>()};
                                   %(RealType)s rcoords[3];
                                   self.dequantize(icoords, rcoords);
                                   return py::make_tuple(rcoords[0], rcoords[1], rcoords[2]);
                               }""")
    def dequantize(self,
                   intcoords = "py::tuple"):
        "Convert int coordinates to real."
        return "py::tuple"

    @PYB11const
    def fillTessellation(self,
                         mesh = "Tessellation<3, %(RealType)s>&"):
        "Read out the current QuantizedTessellation to regular Tessellation."
        return "void"

    #...........................................................................
    # Properties
    xmin = PYB11property("py::tuple",
                         getterraw = """[](const QuantizedTessellation3d<%(IntType)s, %(RealType)s>& self) {
                             return py::make_tuple(self.xmin[0], self.xmin[1]);
                           }""",
                         setterraw = """[](QuantizedTessellation3d<%(IntType)s, %(RealType)s>& self,
                                           py::tuple val) {
                             self.xmin[0] = val[0].cast<%(RealType)s>();
                             self.xmin[1] = val[1].cast<%(RealType)s>();
                           }""",
                         doc = "The minimum (real) coordinate")

    xmax = PYB11property("py::tuple",
                         getterraw = """[](const QuantizedTessellation3d<%(IntType)s, %(RealType)s>& self) {
                             return py::make_tuple(self.xmax[0], self.xmax[1]);
                           }""",
                         setterraw = """[](QuantizedTessellation3d<%(IntType)s, %(RealType)s>& self,
                                           py::tuple val) {
                             self.xmax[0] = val[0].cast<%(RealType)s>();
                             self.xmax[1] = val[1].cast<%(RealType)s>();
                           }""",
                         doc = "The maximum (real) coordinate")

    #...........................................................................
    # Attributes
    length = PYB11readwrite()
    infRadius = PYB11readwrite()
    generators = PYB11readwrite()
    guardGenerators = PYB11readwrite()
    nodes = PYB11readwrite()
    edges = PYB11readwrite()
    faces = PYB11readwrite()
    cellFaces = PYB11readwrite()

#-------------------------------------------------------------------------------
# Template instantiations
QuantizedTessellation3d_KD = PYB11TemplateClass(QuantizedTessellation3d, template_parameters=("polytope::KeyTraits::Key", "double"), pyname="QuantizedTessellation3d")
